aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -799,4 +800,241 @@ Tensor sspaddmm(const Tensor& self, const Tensor& mat1, const Tensor& mat2,
   return result;
 }
 
+// --------------------------------------------------------------------
+// sparse.sum()
+// --------------------------------------------------------------------
+Tensor _sparse_sum(const SparseTensor& input) {
+  return input.values().sum();
+}
+
+Tensor _sparse_sum(const SparseTensor& input, ScalarType dtype) {
+  return at::_sparse_sum(input.to(dtype));
+}
+
+Tensor _sparse_sum(const SparseTensor& input, IntList dims_to_sum) {
+  return at::_sparse_sum(input, dims_to_sum, false);
+}
+
+Tensor _sparse_sum(const SparseTensor& input, IntList dims_to_sum, ScalarType dtype) {
+  return at::_sparse_sum(input.to(dtype), dims_to_sum, false);
+}
+
+Tensor _sparse_sum(const SparseTensor& input, IntList dims_to_sum, bool keepdim) {
+
+  AT_CHECK(input.is_coalesced(), "To support autograd, input SparseTensor has to be coalesced.");
+
+  const int64_t input_dim = input.dim();
+  dim_list_to_bitset(dims_to_sum, input_dim);
+  auto dims_to_sum_v = dims_to_sum.vec();
+  maybe_wrap_dims(dims_to_sum_v, input_dim);
+
+  std::bitset<64> dims_to_sum_b;
+  for (auto d : dims_to_sum_v) {
+    dims_to_sum_b[d] = true;
+  }
+
+  LongTensor indices = input._indices();
+  Tensor values = input._values();
+  IntList sizes = input.sizes();
+  const int64_t sparse_dim = input.sparse_dim();
+  const int64_t dense_dim = input.dense_dim();
+
+  int64_t sparse_dims_to_sum_size = 0;
+  auto dims_to_keep_v = std::vector<int64_t>();
+  Tensor sparse_dims_to_keep_t = at::zeros({sparse_dim}, indices.options().dtype(at::kByte));
+  auto dense_dims_to_sum_v = std::vector<int64_t>();
+  for (int64_t d = 0; d < input_dim; d++) {
+    if (dims_to_sum_b[d]) {
+      if (d < sparse_dim) sparse_dims_to_sum_size ++;
+      else dense_dims_to_sum_v.emplace_back(d + 1 - sparse_dim);
+    }
+    else {
+      dims_to_keep_v.emplace_back(d);
+      if (d < sparse_dim) sparse_dims_to_keep_t[d].fill_(1);
```

Comment:
This is quite slow. And even slower on CUDA. For CPU, you can use the tensor accessor instead, which will be much faster. For CUDA, there is probably a better strategy too. Even creating CPU tensor and sending to CUDA is probably faster (even though you have a synchronization).